Add sigma-delta DAC LiteX core.

tcal-x opened the pull request:
Tested on Arty and Icebreaker.

Sample usage from the BaseSoC for Arty, to output the analog signal on the first pin of PMOD B:
```
    def add_dac(self):
        self.platform.add_extension(arty.raw_pmod_io("pmodb"))
        outsig = self.platform.request("pmodb")[0]
        self.submodules.dac = DAC(outsig, 12)
```

The value to convert is written from software using:
```
#include <generated/csr.h>

    uint32_t v = .....;
    dac_value_write(v);
```




Signed-off-by: Tim Callahan <tcal@google.com>

Commit:
Add sigma-delta DAC LiteX core.

Signed-off-by: Tim Callahan <tcal@google.com>

enjoy-digital commented:
Thanks @tcal-x, I'll have a look and do some tests soon.

Dolu1990 commented:
Hi ^^

I did a similar thing in SpinalHDL, and basicaly, outputing the MSB of the register endup creating quite some distortion.
The reason seems to be related to the switching caracteristics of the IO.

I had much better results (SNR / FFT) using something in that kind  : 
https://github.com/SpinalHDL/SpinalHDL/blob/e48dc0eb13b3d8139645a81ae824014909f2fb59/lib/src/main/scala/spinal/lib/misc/analog/DacSigmaDelta.scala#L23

Where basicaly, depending the value of the MSB, you put a given 3 cycle sequance on the pin :
- MSB = '1' => 1,1,0 on the pin
- MSB = '0' => 1,0,0 on the pin

This ensure that the frequancy of the signal on the pin stay constant, at the cost of a frequancy divided by 3.

enjoy-digital commented:
Thanks @Dolu1990! I'm not yet familiar with Delta-Signal DACs but @tcal-x implementation seemed too simple to be true :)

@tcal-x; Do you want to have a look at implementing it the way @Dolu1990 just described it or do you want me to have a look? (I'm interested to explore this, so could re-implement it if you don't have time).

Dolu1990 commented:
Some ref : 
https://opencores.org/projects/sigma_delta_dac_dual_loop

```
dsm2
allows to obtain higher clock frequency, and therefore higher oversampling ratio, but number of rising and falling slopes in time unit depends on signal value. Therefore you may experience nonlinear distortions if those two slopes are not symmetrical.
```

tcal-x commented:
Thanks for the comments @enjoy-digital and @Dolu1990 !

Some background -- I needed a DAC just to check that I was getting good data from an ADC, specifically the Digilent MIC3 PMOD (https://digilent.com/shop/pmod-mic3-mems-microphone-with-adjustable-gain/).   I wasn't going to use the DAC in the final design.   But it actually sounded better than I expected (the overall connection from MIC3 through the DAC shown here)!   So I thought it might be generally useful if someone wanted a quick and easy monitor on a digital audio data stream.

I got my original design from https://www.fpga4fun.com/PWM_DAC_3.html.   But I think the description there is wrong --- they DON'T have a PWM DAC ---  PWM would lump all of the 1's together for each audio sample, while the sigma delta DAC intermixes the 1's and 0's as much as possible.

Dolu1990, is the better SNR with your approach because the number of transitions stays constant?   Or because the output voltage doesn't approach the bounds?   Or both :)   ?      Were you able to measure the SNR with equipment, or did it sound better to your ear?

Dolu1990 commented:
> Dolu1990, is the better SNR with your approach because the number of transitions stays constant? Or because the output voltage doesn't approach the bounds? Or both :)

I think it is more related to the number of transition staying constant.

> Were you able to measure the SNR with equipment, or did it sound better to your ear?

I do not have any traces of the mesurements i did, but from memory, the sigma delta without the 3 cycle sequance had visible artefact on both the scope and FFT.
A sine wave test i did was very distorted, especialy in the middle range, and on the FFT the amount of armonics was more than 20 db worst.

But that was some fast test with a cheap scope ^^

tcal-x commented:
@enjoy-digital it's fine to close this without merging.   As we've seen there's many different way of building the DAC.   I mostly wanted to make a PR so that I could find the core easily down the road if I needed it.

enjoy-digital commented:
@tcal-x: I'm interested by the core but just want to play a bit with it, do some tests and itnegrate @Dolu1990's improvements.

tcal-x commented:
:+1:

enjoy-digital commented:
Thanks @tcal-x, this has been merged with follow up improvements (taking @Dolu1990 advices into account, improving codestyle, adding a test bench, etc...).

enjoy-digital committed:
soc/cores/dac: Add constant-transition mode

Clean up the sigma-delta DAC core while preserving the original CSR-backed full-rate output as the default behavior.

Add an optional constant-transition mode based on Dolu1990's PR feedback. In this mode, each sigma-delta bit is emitted as a three-cycle symbol: 1,1,0 for one and 1,0,0 for zero. This keeps output transition activity regular at the cost of output rate/range, so it remains opt-in.

Add simulation coverage for both output modes.

enjoy-digital committed:
soc/cores/dac: Follow LiteX core structure

Move the optional CSR handling to an add_csr() helper and allow the output signal to be created internally when no pad/signal is provided.

This keeps the DAC closer to the structure used by simple LiteX cores such as PWM, with public runtime signals declared first, hardware after the separator, and CSR wiring added as an optional integration layer.
